db/migrate/20240109180000_fix_unread_comments_inconsistencies.rb

```diff
@@ -0,0 +1,54 @@
+# Fixes the unread_comments flag for all users. Unintended behavior was
+# introduced in pull request #515. Migration introduced in #587.
+# Behavior fixed in #585.
+#
+# This migration is generally *not* idempotent since users might have interacted
+# with the website since the migration was run and thus they will probably have
+# different unread comments flags as the ones at the time of the migration.
+#
+# This migration is not reversible as we don't store the previous state of
+# the unread_comments flag.
+class FixUnreadCommentsInconsistencies < ActiveRecord::Migration[7.0]
+  def change
+    users = User.all
+    num_fixed_users = 0
+
+    users.find_each do |user|
+      was_user_fixed = fix_unread_comments_flag(user)
+      num_fixed_users += 1 if was_user_fixed
+    end
+
+    Rails.logger.debug { "Ran through #{users.length} users (unread comments flag)" }
+    Rails.logger.debug { "Fixed #{num_fixed_users} users (unread comments flag)" }
+  end
+
+  # Fixes the unread_comments flag for a given user.
+  # Returns true if the flag needed a change (and was changed), false otherwise.
+  def fix_unread_comments_flag(user)
+    readers = Reader.where(user_id: user.id)
+    return false if readers.blank?
+
+    had_user_unread_comments = user.unread_comments
+    has_user_unread_comments = false
+
+    readers.each do |reader|
+      thread = Commontator::Thread.find_by(id: reader.thread_id)
+      next if thread.blank? # thread_id should never be nil, just to be sure
```

Comment:
If `thread_id` is `nil` or some other value that is not attained and no thread is found, then` thread` will bei `nil`. So, the statement 

```ruby
next unless thread
```
should actually work for all possible cases, since nil is falsy and an actual thread is truthy. If you want to make it more verbatim, I would probably rather use 

```ruby
next if thread.nil?
```
instead of `thread.blank?` but that is maybe a personal preference of mine. See also [here](https://stackoverflow.com/questions/885414/how-to-understand-nil-vs-empty-vs-blank-in-ruby).
If you want to leave it as it is this is also fine with me.

Reply:
Changed to `next if thread.nil?` in 3eb6157ecd2.